aleph-client/src/contract/event.rs

```diff
@@ -0,0 +1,220 @@
+//! Utilities for listening for contract events.
+//!
+//! To use the module you will need to first create a subscription (a glorified `Receiver<String>`),
+//! then run the listen loop. You might want to run the loop in a separate thread.
+//!
+//! ```no_run
+//! # use std::sync::mpsc::channel;
+//! # use std::thread;
+//! # use std::time::Duration;
+//! # use aleph_client::{Connection};
+//! # use aleph_client::contract::ContractInstance;
+//! # use aleph_client::contract::event::{listen_contract_events, subscribe_events};
+//! # use anyhow::Result;
+//! # use sp_core::crypto::AccountId32;
+//! # fn example(conn: Connection, address1: AccountId32, address2: AccountId32, path1: &str, path2: &str) -> Result<()> {
+//!     let subscription = subscribe_events(&conn)?;
+//!     let contract1 = ContractInstance::new(address1, path1)?;
+//!     let contract2 = ContractInstance::new(address2, path2)?;
+//!     let (cancel_tx, cancel_rx) = channel();
+//!
+//!     thread::spawn(move || {
+//!         listen_contract_events(
+//!             subscription,
+//!             &[&contract1, &contract2],
```

Comment:
I find it a bit inconvenient to 'lose' my contract instance for just event listening. Imagine a scenario, where I want to:
 1. execute some contract action
 2. wait for a confirmation event
 3. execute another contract action (for the same contract)
 4. wait for a confirmation event

With the current API I have to create 4 contract instances, because spawning listening requires ownership of the contract instance. I see that `InkProject` doesn't implement `Clone`, but I think we could capture and save the metadata path in `ContractInstance` and expose a custom cloning method.

Reply:
We can't avoid passing ownership of `ContractInstance` since we need `InkProject` to decode `Raw` into `ContractEvent`. But what we can do is to split the interface provided by this module so that (note that `decode_contract_event` does two below things actually)
* we get all `Raw` events filtered by contract account id , ie `Vec<Raw>` 
* and then we apply conversion on that vector, in the same thread
Then for listening to itself we just need to pass AccountId
The disadvantage of the above proposal is that current version is more concise

Reply:
You don't have to lose it! I made `listen` take `&[&ContractInstance]` on purpose. It's a bit fiddly, but what you can do is pass the contracts to the listening thread via an `Arc` and then extract `&ContractInstance` from the `Arc` for listening.

Perhaps I should add this to the example?

Reply:
yes, definitely, please